fix(TreeList): fixes + documentation (#1447)
gravity-ui/uikit: src/components/TreeList/__stories__/TreeList.mdx

    {data: {id: 'id-2', title: 'some title 2'}, children: [...]},
];

<TreeList items={items} getId={({id})=> id}  />
```

### qa

Set `qa` attribute for the container and sheet elements. `qa` attribute is also passed to the `ListItemView`.

> Use the [getListItemQa](/docs/unstable-uselist--docs#getlistitemqa) is used to generate `qa` attributes in list items; also use this function in tests to compute a unique data attribute to access a specific list item

    ```ts
    await locator.getByTestId(getListItemQa('some-list-qa', '0')); // select the first item in the list if auto-generated IDs are not used
    ```

### className

Pass custom CSS class for the list container.

### id

Set a custom id data attribute. By default, a unique identifier will be assigned.

### multiple

This prop is necessary for the correct view of the selected elements since the state of the selected elements is controlled from the parent component.

### setActiveItemId

Required for correct keyboard interactions. While navigating through the keyboard, you need to set the next active element.

### defaultGroupsExpanded

Control the default expanded state of items' groups. Default - `true`.

### renderItem

> The ability to completely redefine the rendering of a list item. For example, add dividers between list items or wrap an item in a link component. As a view component to display a list item, use [ListItemView](/docs/unstable-uselist--docs#listitemview);

Reviewer: ```suggestion
Redefine the rendering of a list item. For example, add dividers between list items or wrap an item in a link component. As a view component to display a list item, use [ListItemView](/docs/unstable-uselist--docs#listitemview);
```

    }}
/>
```

#### renderItem function argument object:

-   `data` - access to the original object with the data of the sheet element;
-   `props` - default props generated by the component taking into account the state (whether the element is selected or not, active, disclosed). The set of returned passes corresponds to the result of the function execution [getItemRenderState](/docs/unstable-uselist--docs#item-state-props);
-   `index` - ordinal index of the element, taking into account that with a tree-like data structure, the list elements have a flatten representation;
-   `renderContainerProps` - the passes thrown from the redefined container. In the vast majority of cases, you won't need it, but it's worth knowing that there is such a possibility.;
-   `context` - useful information about the current list item:
    -   `itemState` - meta info about item
        -   `indentation` - integer number representing nested list level
        -   `parentId` - `id` of parent list item;
    -   `groupState` - An optional parameter. If the list item is also the first item of the nested list:
        -   `childrenIds` - array of `id` of nested list items;

>     -   `isLastItem` - is the current item the last one in the list;

Reviewer: Let's document this like a TS interface:

**RenderItemProps**
```ts
interface RenderItemProps {
    // access to the original object with the data of the sheet element
    data: any
    // ...
}
```

> Important! Absolutely all the props for [ListItemView](/docs/unstable-uselist--docs#listitemview) can be redefined in the renderItem method. This is the preferred method for changing the view of the sheet elements.

### renderContainer

> Ability to override default list container `ListContainerView`;

Reviewer: ```suggestion
Render custom list container.
```

```tsx
<TreeList
    renderContainer={({
        id,
        containerRef,
        visibleFlattenIds,
        renderItem,
        size,
        className,
        qa,
    }: TreeListRenderContainerProps<T>) => {
        return (
            <ListContainerView
                qa={qa}
                className={className}

>                 fixedHeight // by default 300px. Use `--g-list-height` css custom propery to override container height if this prop active

Reviewer: I just noticed. Why we are using CSS here instead of having `height: number` prop?

Author: it is a legacy from previous list. We can improve this moment later

                id={id}
                ref={containerRef}
            >
                <VirtualizedListContainer // your own implementation based on the choice of a library for virtualization
                    items={visibleFlattenIds}
                    itemSize={(_index) => computeItemSize(size)}
                >
                    {(id, index) =>
                        renderItem(
                            id,
                            index,
                            _, // here you can optionally pass any props depending of render context */,
                        )
                    }
                </VirtualizedListContainer>
            </ListContainerView>
        );
    }}
/>
```

### onItemClick

> Ability to define on item click callback. Also this callback will be called with handling keyboard actions

Reviewer: ```suggestion
Item's click callback. It also will be called on keyboard actions.
```

```tsx
<TreeList
    onItemClick={({
        id,
        index,
        selected,
        disabled,
        expanded,
        context: {
            itemState: {indentation, parentId},
            groupState: {childrenIds},
            isLastItem,
        },
    }) => {
        // just do it!
    }}
/>
```

> #### onItemClick function argument object:

Reviewer: Same as for `RenderItemProps`

import {Meta} from '@storybook/addon-docs';

<Meta title="Unstable/TreeList" />

# TreeList

The basic component for working with lists, including tree-like ones. Under the hood, it uses the [useList](/docs/unstable-uselist--docs). To manage the state, it is recommended to use the [useListState](/docs/unstable-uselist--docs#useliststate) hook.

`Storybook` provides complex examples how to use this components from this documentation.

## Props:

-   [items](#items);
-   [mapItemDataToProps](#mapitemdatatoprops);
-   [qa](#qa);
-   [id](#id);
-   [containerRef](#containerref);
-   [className](#classname);
-   [multiple](#multiple);
-   [size](#size-available-options);
-   [defaultGroupsExpanded](#defaultgroupsexpanded);
-   [getId](#getid);
-   [renderItem](#renderitem);
-   [renderContainer](#rendercontainer);
-   [onItemClick](#onitemclick);
-   [...useListState](/docs/unstable-uselist--docs#useliststate)

## Quick start:

### Basic example:

```tsx

> import {type ListItemType, getItemRenderState, TreeList} from '@gravity-ui/uikit';

Reviewer: `getItemRenderState ` not used in the example

const items: ListItemType<string>[] = ['one', 'two', 'free', 'four', 'five'];

> <TreeList items={items} mapItemDataToProps={(title) => ({title})} />;

Reviewer: I would name the parameter as `item` to emphasize that it is the element of items array

```

### Example with state:

```tsx

> import {type ListItemType, getItemRenderState, TreeList} from '@gravity-ui/uikit';

Reviewer: `getItemRenderState ` not used in the example

const items: ListItemType<string>[] = [
    {title: 'one'},
    {title: 'two'},
    {title: 'free'},
    {title: 'four'},
    {title: 'five'},
];

const Component = () => {
    const listState = useListState();

    const handleItemClick: TreeListOnItemClick<string> = ({id, disabled, groupState}) => {
        if (disabled) return;

        if (groupState) {
            listState.setExpanded((prevState) => ({
                ...prevState,
                [id]: id in prevState ? !prevState[id] : false,
            }));
        } else {
            listState.setSelected((prevState) => ({
                [id]: !prevState[id],
            }));
        }

        listState.setActiveItemId(id);
    };

    return (
        <TreeList
            items={items}
            {...listState}
            onItemClick={handleItemClick}
            mapItemDataToProps={({title}) => ({title})}
        />
    );
};
```

> > If you want to display the nodes of the list as regular elements without the possibility of hiding the folded elements of the sheet, then just do not pass the `expandedById` object from the tate to the component itself:

Reviewer: ```suggestion
> If you want to display the nodes of the list as regular elements without the possibility of hiding the folded elements of the sheet, then just do not pass the `expandedById` object from the state to the component itself:
```

```ts
const {expandedById, setExpandedById, ...listState} = useListState();

<TreeList {...listSttate} />
```

## Component props:

### items

> array of list item. More details about data structure and properties you can find [here](/docs/unstable-uselist--docs#items-supported-data-structure);

Reviewer: ```suggestion
Array of list items. More details about data structure and properties you can find [here](/docs/unstable-uselist--docs#items-supported-data-structure);
```

### mapItemDataToProps

> map list item data structire to `ListItemView` [props](/docs/unstable-uselist--docs#listitemview);

Reviewer: ```suggestion
Map list item data structire to `ListItemView` [props](/docs/unstable-uselist--docs#listitemview);
```

### containerRef

> The ability to pass a link to the DOM element of the container. For example, in order to control the focus of the list to activate keyboard navigation support;

Reviewer: ```suggestion
Pass a ref to the DOM element of the container. For example, in order to control the focus of the list to activate keyboard navigation support;
```

```tsx
import React from 'react';

> import {type ListItemType, getItemRenderState, TreeList, Button, Alert} from '@gravity-ui/uikit';

Reviewer: `getItemRenderState ` not used in the example

const items: ListItemType<string>[] = [
    {data: {title: 'one'}},
    {data: {title: 'two'}},
    {data: {title: 'free'}},
    {data: {title: 'four'}},
    {data: {title: 'five'}},
];

const Component = () => {
    const containerRef = React.useRef<HTMLDivElement>(null);
    const listState = useListState();

    const handleItemClick: TreeListOnItemClick<string> = ({
        id,
        disabled,
        selected,
        expanded,
        groupState,
    }) => {
        // ...
    };

    return (
        <>
            <Alert

>                 message="by button click. After click you can navigate list during keyboard"

Reviewer: ```suggestion
                message="After clicking the button you can navigate list during keyboard"
```

                theme="info"
            />
            <Button
                onClick={() => {
                    containerRef?.current?.focus();
                }}
            >
                focus list
            </Button>
            <TreeList
                items={items}
                {...listState}
                containerRef={containerRef}
                onItemClick={handleItemClick}
                mapItemDataToProps={({title}) => ({title})}
            />
        </>
    );
};
```

### getId

> Ability to generate an id for a list item depending on the list data. It is necessary to have access to more custom management of the state of the list. The property is optional.

Reviewer: ```suggestion
Generate an id for a list item depending on the list's data. If it's necessary to have access for custom management of the state of the list. The property is optional.
```

```tsx
const items = [
    {data: {id: 'id-1', title: 'some title 1'}, children: [...]},
    {data: {id: 'id-2', title: 'some title 2'}, children: [...]},
];

<TreeList items={items} getId={({id})=> id}  />
```

### qa

> The ability to set a qa attribute for the container and sheet elements. The Qa attribute is also passed to the ListItemView.

Reviewer: Here and below `The ability to ` feels redundant to say

```suggestion
Set  `qa` attribute for the container and sheet elements. `qa` attribute is also passed to the `ListItemView.`
```

> Use the [getListItemQa](/docs/unstable-uselist--docs#getlistitemqa) is used to generate `qa` attributes in list items; also use this function in tests to compute a unique data attribute to access a specific list item

    ```ts
    await locator.getByTestId(getListItemQa('some-list-qa', '0')); // select the first item in the list if auto-generated IDs are not used
    ```

### className

> The ability to transfer custom css class name for the list container;

Reviewer: ```suggestion
Pass custom CSS class for the list container.
```

### id

> The ability to set a custom id data attribute. By default, a unique identifier will be assigned;

Reviewer: ```suggestion
Set a custom id data attribute. By default, a unique identifier will be assigned.
```

### multiple

> Since the state of the selected elements is controlled from above the component, this prop is necessary for the correct visual display of the selected elements;

Reviewer: ```suggestion
This prop is necessary for the correct view of the selected elements since the state of the selected elements is controlled from the parent component.
```

### setActiveItemId

> > Required for keyboard correct work. Because when navigating from the keyboard, you need to set the next active element

Reviewer: Why quote?

> > Required for keyboard correct work. Because when navigating from the keyboard, you need to set the next active element

Reviewer: ```suggestion
Required for correct keyboard interactions. While navigating through the keyboard, you need to set the next active element.
```

### defaultGroupsExpanded

> default value - true
> 
> Ability to handle default groups expanded behavior.
> Works if `expandedById` state passed

Reviewer: ```suggestion
Control the default expanded state of items' groups. Default - `true`.
```
